Working AES-CBC using the CAU on the Teensy 3

jeremy-daily opened the pull request.

jeremydaily committed:
Working AES-CBC using the CAU on the Teensy 3

BenGardiner commented on tests/cryptolibAESSHA/cau_api.h:
```diff
@@ -0,0 +1,390 @@
+/*
```
is this *your* header file? did you import it from somewhere else? If so, how is it licensed?

jeremy-daily replied:
This is from Freescale. I've included the details on the license from their download.

BenGardiner commented on tests/cryptolibAESSHA/cryptolibAESSHA.ino:
```diff
@@ -0,0 +1,137 @@
+    for (uint8_t i = 0; i < 16; i++){
+      in[i] = data[j+i] ^ out[i];
+    }
+    cau_aes_encrypt (in, keysched, 10, out); // # 16-byte block
```
magic number `10` -- I recommend defining this as a descriptive-named constant. e.g. AES_128_NROUNDS

jeremy-daily replied:
Done.

BenGardiner commented on tests/cryptolibAESSHA/cryptolibAESSHA.ino:
```diff
@@ -0,0 +1,137 @@
+  
+  t = micros();
+  memcpy(iv,init_vector,16);
+  aes_cbc_decrypt(clear_text, cipher_text);
```
you're testing 'round-trip' through your aes-cbc routines; this is a good test.

It is necessary but not sufficient. You should also test with known test vectors. e.g. https://github.com/google/wycheproof/tree/master/testvectors

jeremy-daily replied:
Looks interesting. I'll have to look into this more. It looks challenging to port...

BenGardiner replied:
No need to port the project; but you should include tests that confirm that your implementation here is capabale of encrypting a known plaintext into and expected ciphertext and the analogous test for decrypt.  Right now your tests are only confirming that you have a pair of functions that are a matching encrypt/decrypt; but you're not confirming that existing tools/implementations could encrypt or decrypt the outputs of your implementations

jeremy-daily replied:
We tested with the online calculator at https://www.cryptogrium.com/aes-cbc.html and obtained matching results.

BenGardiner replied:
You should use the encrypted output of that online calculator as the input in a test here (and vice versa). Right now your tests verify that your implementation can do a round trip. You are missing a test that to confirm that the implementation will 'work with standard tools'.

BenGardiner commented on tests/cryptolibAESSHA/mmcau_des_functions.S:
```diff
@@ -0,0 +1,257 @@
+#*******************************************************************************
+#*******************************************************************************
+#
+# Copyright (c) Freescale Semiconductor, Inc 2011.
```
under what license does Freescale distribute this assembler file?

BenGardiner replied:
same question for the remaining `.S` files too

jeremy-daily replied:
Posted the license. It's permissive.

BenGardiner commented on tests/cryptolibAESSHA/mmcau_aes_functions.S:
```diff
@@ -0,0 +1,929 @@
+#*******************************************************************************
+#*******************************************************************************
+#
+# Copyright (c) Freescale Semiconductor, Inc 2011.
```
under what license does Freescale distribute this assembler file?

jeremy-daily replied:
Posted the license. It's permissive.

jeremydaily committed:
Added Readmes and licenses

jeremydaily committed:
Removed Magic Numbers and included hdr

BenGardiner commented on tests/cryptolibAESSHA/cryptolibAESSHA.ino:
```diff
@@ -4,7 +4,9 @@ Uses 512 Byte block examples
 #include "cau_api.h"
+
 #define data_size 512
+#define AES_ROUNDS 10
```
not quite descriptive enough; This is a *value* for the number of AES rounds, true; however, for AES other than AES-128 it is not a valid number of rounds...

jeremy-daily replied:
Easy change. I replaced the one @dcn1470 entered too.

duyvan1995 committed:
Add AES CBC function for CAN Log

Use CAU on the K66 to encrypt log data, key and IV are randomly generated and stored in baudrate.txt for testing purposes only

jeremydaily committed:
Changed AES_128_NROUNDS

duyvan1995 committed:
Fix Typo Error

BenGardiner commented:
I still think you should have tests against static/constant/literal test-vectors and not just round-trip your own implementations. But PR #6 is on-top-of this one and I want to review it separately.

jeremy-daily commented:
Agreed. I think we should start a new branch with the test vectors.
